File: sdks/python/prsm_sdk/ftns.py
"""
PRSM SDK FTNS Token Manager
Handles FTNS token operations including balance, transfers, and staking
"""

import structlog
from typing import Optional, List, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field

from .exceptions import InsufficientFundsError, PRSMError

logger = structlog.get_logger(__name__)


class FTNSBalance(BaseModel):
    """FTNS token balance information"""
    total_balance: float = Field(..., description="Total FTNS balance")
    available_balance: float = Field(..., description="Available for spending")
    reserved_balance: float = Field(..., description="Reserved for pending operations")
    staked_balance: float = Field(0.0, description="Staked for network participation")
    earned_today: float = Field(0.0, description="FTNS earned today")
    spent_today: float = Field(0.0, description="FTNS spent today")
    last_updated: datetime = Field(..., description="Last balance update")


class Transaction(BaseModel):
    """FTNS transaction record"""
    transaction_id: str = Field(..., description="Unique transaction ID")
    transaction_type: str = Field(..., description="Type: transfer, stake, unstake, reward")
    amount: float = Field(..., description="Transaction amount in FTNS")
    from_address: Optional[str] = Field(None, description="Sender address")
    to_address: Optional[str] = Field(None, description="Recipient address")
    status: str = Field(..., description="Status: pending, completed, failed")
    timestamp: datetime = Field(..., description="Transaction timestamp")
    memo: Optional[str] = Field(None, description="Transaction memo")
    block_number: Optional[int] = Field(None, description="Block number")


class StakeInfo(BaseModel):
    """Staking information"""
    staked_amount: float = Field(..., description="Currently staked amount")
    rewards_earned: float = Field(..., description="Total rewards earned")
    lock_period: int = Field(..., description="Lock period in days")
    apy: float = Field(..., description="Annual percentage yield")
    unlock_date: Optional[datetime] = Field(None, description="When stake unlocks")


class TransferRequest(BaseModel):
    """Transfer request parameters"""
    to_address: str = Field(..., description="Recipient address")
    amount: float = Field(..., gt=0, description="Amount to transfer")
    memo: Optional[str] = Field(None, description="Optional memo")


class TransferResponse(BaseModel):
    """Transfer response"""
    transaction_id: str = Field(..., description="Transaction ID")
    status: str = Field(..., description="Transaction status")
    amount: float = Field(..., description="Transferred amount")
    fee: float = Field(..., description="Transaction fee")
    timestamp: datetime = Field(..., description="Transaction timestamp")


class FTNSManager:
    """
    Manages FTNS token operations
    
    Provides methods for:
    - Checking balances
    - Transferring tokens
    - Staking/unstaking
    - Transaction history
    """
    
    def __init__(self, client):
        """
        Initialize FTNS manager
        
        Args:
            client: PRSMClient instance for making API requests
        """
        self._client = client
    
    async def get_balance(self) -> FTNSBalance:
        """
        Get current FTNS balance
        
        Returns:
            FTNSBalance with current balance information
            
        Example:
            balance = await client.ftns.get_balance()
            print(f"Available: {balance.available_balance} FTNS")
        """
        response = await self._client._request("GET", "/ftns/balance")
        return FTNSBalance(**response)
    
    async def transfer(
        self,
        to_address: str,
        amount: float,
        memo: Optional[str] = None
    ) -> TransferResponse:
        """
        Transfer FTNS tokens to another address
        
        Args:
            to_address: Recipient address
            amount: Amount to transfer
            memo: Optional memo for the transaction
            
        Returns:
            TransferResponse with transaction details
            
        Raises:
            InsufficientFundsError: If balance is insufficient
            
        Example:
            result = await client.ftns.transfer(
                to_address="0x123...",
                amount=10.5,
                memo="Payment for services"
            )
        """
        # Check balance first
        balance = await self.get_balance()
        if balance.available_balance < amount:
            raise InsufficientFundsError(amount, balance.available_balance)
        
        request = TransferRequest(
            to_address=to_address,
            amount=amount,
            memo=memo
        )
        
        response = await self._client._request(
            "POST",
            "/ftns/transfer",
            json_data=request.model_dump()
        )
        
        return TransferResponse(**response)
    
    async def get_history(
        self,
        limit: int = 50,
        offset: int = 0,
        transaction_type: Optional[str] = None
    ) -> List[Transaction]:
        """
        Get transaction history
        
        Args:
            limit: Maximum number of transactions to return
            offset: Offset for pagination
            transaction_type: Filter by type (transfer, stake, unstake, reward)
            
        Returns:
            List of Transaction objects
            
        Example:
            history = await client.ftns.get_history(limit=10)
            for tx in history:
                print(f"{tx.transaction_type}: {tx.amount} FTNS")
        """
        params = {"limit": limit, "offset": offset}
        if transaction_type:
            params["type"] = transaction_type
        
        response = await self._client._request(
            "GET",
            "/ftns/history",
            params=params
        )
        
        return [Transaction(**tx) for tx in response.get("transactions", [])]
    
    async def stake(self, amount: float, lock_period: int = 30) -> StakeInfo:
        """
        Stake FTNS tokens for network participation
        
        Args:
            amount: Amount to stake
            lock_period: Lock period in days (default: 30)
            
        Returns:
            StakeInfo with staking details
            
        Raises:
            InsufficientFundsError: If balance is insufficient
            
        Example:
            stake_info = await client.ftns.stake(100, lock_period=60)
            print(f"Staked {amount} FTNS at {stake_info.apy}% APY")
        """
        balance = await self.get_balance()
        if balance.available_balance < amount:
            raise InsufficientFundsError(amount, balance.available_balance)
        
        response = await self._client._request(
            "POST",
            "/ftns/stake",
            json_data={"amount": amount, "lock_period": lock_period}
        )
        
        return StakeInfo(**response)
    
    async def unstake(self, amount: Optional[float] = None) -> StakeInfo:
        """
        Unstake FTNS tokens
        
        Args:
            amount: Amount to unstake (if None, unstake all)
            
        Returns:
            StakeInfo with updated staking details
            
        Example:
            stake_info = await client.ftns.unstake(50)
        """
        json_data = {}
        if amount is not None:
            json_data["amount"] = amount
        
        response = await self._client._request(
            "POST",
            "/ftns/unstake",
            json_data=json_data
        )
        
        return StakeInfo(**response)
    
    async def get_stake_info(self) -> StakeInfo:
        """
        Get current staking information
        
        Returns:
            StakeInfo with current staking details
        """
        response = await self._client._request("GET", "/ftns/stake/info")
        return StakeInfo(**response)
    
    async def estimate_transfer_fee(self, amount: float) -> float:
        """
        Estimate the fee for a transfer
        
        Args:
            amount: Transfer amount
            
        Returns:
            Estimated fee in FTNS
        """
        response = await self._client._request(
            "POST",
            "/ftns/estimate-fee",
            json_data={"amount": amount}
        )
        return response.get("fee", 0.0)

File: sdks/go/ftns/ftns.go
// Package ftns provides FTNS token management for the PRSM Go SDK
package ftns

import (
	"context"
	"time"

	"github.com/pkg/errors"
)

// Manager handles FTNS token operations
type Manager struct {
	client HTTPClient
}

// HTTPClient interface for making HTTP requests
type HTTPClient interface {
	MakeRequest(ctx context.Context, method, endpoint string, reqBody interface{}, respBody interface{}) error
}

// New creates a new FTNS manager
func New(client HTTPClient) *Manager {
	return &Manager{
		client: client,
	}
}

// FTNSBalance represents FTNS token balance information
type FTNSBalance struct {
	TotalBalance     float64   `json:"total_balance"`
	AvailableBalance float64   `json:"available_balance"`
	ReservedBalance  float64   `json:"reserved_balance"`
	StakedBalance    float64   `json:"staked_balance"`
	EarnedToday      float64   `json:"earned_today"`
	SpentToday       float64   `json:"spent_today"`
	LastUpdated      time.Time `json:"last_updated"`
}

// Transaction represents an FTNS transaction record
type Transaction struct {
	TransactionID   string     `json:"transaction_id"`
	TransactionType string     `json:"transaction_type"` // transfer, stake, unstake, reward
	Amount          float64    `json:"amount"`
	FromAddress     string     `json:"from_address,omitempty"`
	ToAddress       string     `json:"to_address,omitempty"`
	Status          string     `json:"status"` // pending, completed, failed
	Timestamp       time.Time  `json:"timestamp"`
	Memo            string     `json:"memo,omitempty"`
	BlockNumber     *int64     `json:"block_number,omitempty"`
}

// StakeInfo represents staking information
type StakeInfo struct {
	StakedAmount  float64    `json:"staked_amount"`
	RewardsEarned float64    `json:"rewards_earned"`
	LockPeriod    int        `json:"lock_period"` // days
	APY           float64    `json:"apy"`
	UnlockDate    *time.Time `json:"unlock_date,omitempty"`
}

// TransferRequest represents a transfer request
type TransferRequest struct {
	ToAddress string  `json:"to_address"`
	Amount    float64 `json:"amount"`
	Memo      string  `json:"memo,omitempty"`
}

// TransferResponse represents a transfer response
type TransferResponse struct {
	TransactionID string    `json:"transaction_id"`
	Status        string    `json:"status"`
	Amount        float64   `json:"amount"`
	Fee           float64   `json:"fee"`
	Timestamp     time.Time `json:"timestamp"`
}

// TransactionHistoryRequest represents parameters for transaction history
type TransactionHistoryRequest struct {
	Limit           int    `json:"limit,omitempty"`
	Offset          int    `json:"offset,omitempty"`
	TransactionType string `json:"transaction_type,omitempty"`
}

// TransactionHistoryResponse represents transaction history response
type TransactionHistoryResponse struct {
	Transactions []Transaction `json:"transactions"`
	Total        int           `json:"total"`
	Offset       int           `json:"offset"`
	Limit        int           `json:"limit"`
}

// StakeRequest represents a staking request
type StakeRequest struct {
	Amount     float64 `json:"amount"`
	LockPeriod int     `json:"lock_period"` // days
}

// GetBalance retrieves the current FTNS balance
func (m *Manager) GetBalance(ctx context.Context) (*FTNSBalance, error) {
	var balance FTNSBalance
	err := m.client.MakeRequest(ctx, "GET", "/api/v1/ftns/balance", nil, &balance)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get FTNS balance")
	}
	return &balance, nil
}

// Transfer transfers FTNS tokens to another address
func (m *Manager) Transfer(ctx context.Context, req *TransferRequest) (*TransferResponse, error) {
	if req == nil {
		return nil, errors.New("transfer request cannot be nil")
	}
	if req.ToAddress == "" {
		return nil, errors.New("to_address is required")
	}
	if req.Amount <= 0 {
		return nil, errors.New("amount must be positive")
	}

	var response TransferResponse
	err := m.client.MakeRequest(ctx, "POST", "/api/v1/ftns/transfer", req, &response)
	if err != nil {
		return nil, errors.Wrap(err, "failed to transfer FTNS")
	}
	return &response, nil
}

// GetTransactionHistory retrieves transaction history
func (m *Manager) GetTransactionHistory(ctx context.Context, req *TransactionHistoryRequest) (*TransactionHistoryResponse, error) {
	if req == nil {
		req = &TransactionHistoryRequest{Limit: 50}
	}
	if req.Limit <= 0 {
		req.Limit = 50
	}

	var response TransactionHistoryResponse
	err := m.client.MakeRequest(ctx, "GET", "/api/v1/ftns/history", req, &response)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get transaction history")
	}
	return &response, nil
}

// Stake stakes FTNS tokens for network participation
func (m *Manager) Stake(ctx context.Context, req *StakeRequest) (*StakeInfo, error) {
	if req == nil {
		return nil, errors.New("stake request cannot be nil")
	}
	if req.Amount <= 0 {
		return nil, errors.New("amount must be positive")
	}
	if req.LockPeriod <= 0 {
		req.LockPeriod = 30 // Default to 30 days
	}

	var response StakeInfo
	err := m.client.MakeRequest(ctx, "POST", "/api/v1/ftns/stake", req, &response)
	if err != nil {
		return nil, errors.Wrap(err, "failed to stake FTNS")
	}
	return &response, nil
}

// Unstake unstakes FTNS tokens
func (m *Manager) Unstake(ctx context.Context, amount *float64) (*StakeInfo, error) {
	req := map[string]interface{}{}
	if amount != nil {
		req["amount"] = *amount
	}

	var response StakeInfo
	err := m.client.MakeRequest(ctx, "POST", "/api/v1/ftns/unstake", req, &response)
	if err != nil {
		return nil, errors.Wrap(err, "failed to unstake FTNS")
	}
	return &response, nil
}

// GetStakeInfo retrieves current staking information
func (m *Manager) GetStakeInfo(ctx context.Context) (*StakeInfo, error) {
	var response StakeInfo
	err := m.client.MakeRequest(ctx, "GET", "/api/v1/ftns/stake/info", nil, &response)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get stake info")
	}
	return &response, nil
}

// EstimateTransferFee estimates the fee for a transfer
func (m *Manager) EstimateTransferFee(ctx context.Context, amount float64) (float64, error) {
	req := map[string]interface{}{"amount": amount}

	var response struct {
		Fee float64 `json:"fee"`
	}
	err := m.client.MakeRequest(ctx, "POST", "/api/v1/ftns/estimate-fee", req, &response)
	if err != nil {
		return 0, errors.Wrap(err, "failed to estimate transfer fee")
	}
	return response.Fee, nil
}
